packages/puppeteer-core/src/bidi/CDPSession.ts

```diff
@@ -0,0 +1,95 @@
+/**
+ * @license
+ * Copyright 2024 Google Inc.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import type ProtocolMapping from 'devtools-protocol/types/protocol-mapping.js';
+
+import {CDPSession} from '../api/CDPSession.js';
+import type {Connection as CdpConnection} from '../cdp/Connection.js';
+import {TargetCloseError, UnsupportedOperation} from '../common/Errors.js';
+import {Deferred} from '../util/Deferred.js';
+
+import type {BrowsingContext} from './BrowsingContext.js';
+
+/**
+ * @internal
+ */
+
+export const cdpSessions = new Map<string, BidiCdpSession>();
+
+/**
+ * @internal
+ */
+export class BidiCdpSession extends CDPSession {
+  #context: BrowsingContext;
+  #sessionId = Deferred.create<string>();
+  #detached = false;
+
+  constructor(context: BrowsingContext, sessionId?: string) {
+    super();
+    this.#context = context;
+    if (!this.#context.supportsCdp()) {
+      return;
+    }
+    if (sessionId) {
+      this.#sessionId.resolve(sessionId);
+      cdpSessions.set(sessionId, this);
+    } else {
+      context.connection
+        .send('cdp.getSession', {
+          context: context.id,
+        })
+        .then(session => {
+          this.#sessionId.resolve(session.result.session!);
+          cdpSessions.set(session.result.session!, this);
+        })
+        .catch(err => {
+          this.#sessionId.reject(err);
+        });
+    }
+  }
+
+  override connection(): CdpConnection | undefined {
+    return undefined;
+  }
+
+  override async send<T extends keyof ProtocolMapping.Commands>(
+    method: T,
+    ...paramArgs: ProtocolMapping.Commands[T]['paramsType']
```

Comment:
Nit: Just to fix the correct type from the base class if its a no-op change.
```suggestion
    params?: ProtocolMapping.Commands[T]['paramsType'][0],
```